### PR TITLE
fix(Wasm): Changing TextBox.Text in code should reset the caret to the beginning

### LINKED ISSUE #7016: Changing `TextBox.Text` in code should reset the caret to the beginning
### Current behavior

See https://github.com/unoplatform/uno/pull/6995#issuecomment-916392149 and see modified tests in the PR.

### Expected behavior

Explained above.

### How to reproduce it (as minimally and precisely as possible)

_No response_

### Workaround

_No response_

### Works on UWP/WinUI

Yes

### Environment

_No response_

### NuGet package version(s)

_No response_

### Affected platforms

WebAssembly

### IDE

_No response_

### IDE version

_No response_

### Relevant plugins

_No response_

### Anything else we need to know?

_No response_

### PR DESCRIPTION
**Pull request by aesteves900**
GitHub Issue (If applicable): closes #7016

## PR Type

What kind of change does this PR introduce?

- Bugfix

## What is the current behavior?

[Issue](https://github.com/unoplatform/uno/pull/6995#issuecomment-916392149)

## What is the new behavior?

When the developer set an invalid range in SelectionStart and SelectionLength the caret reset to the beginning

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ x] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ x] Validated PR `Screenshots Compare Test Run` results.
- [ x] Contains **NO** breaking changes
- [ x] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ x] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.cs [hunk 1]
```diff
@@ -886,7 +904,24 @@ private void DeleteButtonClick()
...
+				}
+				else
+				{
+					SelectionChanged?.Invoke(this, new RoutedEventArgs(this));
+				}
+			}
+			else
+			{
+				SelectionChanged?.Invoke(this, new RoutedEventArgs(this));
+			}
```
Comment (carldebilly) [inline src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.cs:923]: Avoid processing the _negative_ clause first...  it can easily confuse the reading of the code.

```suggestion
			if(Text is { Length: > 0 } text || text.TryGetStringByRange(SelectionStart, SelectionLength, out _))
			{
				SelectionChanged?.Invoke(this, new RoutedEventArgs(this));
			}
			else
			{
				ResetCarret();
			}
```

File: src/Uno.Foundation/Uno.Core.Extensions/Uno.Core.Extensions/StringExtensions.cs [hunk 2]
```diff
@@ -34,6 +34,22 @@ internal static partial class StringExtensions
...
+			try
+			{
+				result = instance.Substring(start, len);
+
+				return true;	
+			}
+			catch
+			{
+				return false;
+			}
```
[the remark this answers is not in the document]
Comment (aesteves900) [reply src/Uno.Foundation/Uno.Core.Extensions/Uno.Core.Extensions/StringExtensions.cs]: It was resolved, thank you!

File: src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.cs (hunk 1 above)
Comment (aesteves900) [reply src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.cs]: The suggestion was commited.

File: src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.cs [hunk 3]
```diff
@@ -886,7 +904,21 @@ private void DeleteButtonClick()
...
+		{
+#if __WASM__
+			if (Text is { Length: > 0 } text || Text.TryGetStringByRange(SelectionStart, SelectionLength, out _))
+			{
+				SelectionChanged?.Invoke(this, new RoutedEventArgs(this));
+			}
+			else
+			{
+				this.ResetCarret();
+			}
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.cs:917]: It'd be best to move this code to the TextBox.wasm.cs file, and use a partial method to call it.

Comment (aesteves900) [reply src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.cs]: It was fixed.

File: src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.cs [hunk 4]
```diff
@@ -374,7 +390,7 @@ protected override void OnFontWeightChanged(FontWeight oldValue, FontWeight newV
 
 		partial void OnForegroundColorChangedPartial(Brush newValue);
 
-		#region PlaceholderText DependencyProperty
+#region PlaceholderText DependencyProperty
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.cs:393]: While useful, avoid mixing cosmetic changes with your feature PR. You can make another PR for this, though.

File: src/Uno.Foundation/Uno.Core.Extensions/Uno.Core.Extensions/StringExtensions.cs [hunk 5]
```diff
@@ -34,6 +34,20 @@ internal static partial class StringExtensions
...
 #endif
 
+		public static bool TryGetStringByRange(this string instance, int start, int len, out string result)
+		{
+			result = null;
+			
+			if (!string.IsNullOrEmpty(instance) && instance.Skip(start).Take(len).ToArray().Length == 0)
+			{
+				return false;
+			}
```
Comment (Youssef1313) [inline src/Uno.Foundation/Uno.Core.Extensions/Uno.Core.Extensions/StringExtensions.cs:44]: Will this have the same meaning?

```suggestion
			if (start + len > instance.Length)
			{
				return false;
			}
```

File: src/Uno.Foundation/Uno.Core.Extensions/Uno.Core.Extensions/StringExtensions.cs [hunk 6]
```diff
@@ -34,6 +34,20 @@ internal static partial class StringExtensions
 		private static readonly Lazy<Regex> _newLineRegex = new Lazy<Regex>(() => new Regex(@"^", RegexOptions.Multiline));
 #endif
 
+		public static bool TryGetStringByRange(this string instance, int start, int len, out string result)
+		{
+			result = null;
+			
+			if (!string.IsNullOrEmpty(instance) && instance.Skip(start).Take(len).ToArray().Length == 0)
```
Comment (Youssef1313) [inline src/Uno.Foundation/Uno.Core.Extensions/Uno.Core.Extensions/StringExtensions.cs:41]: If `instance` is null, this will throw an `ArgumentNullException`.

`!string.IsNullOrEmpty(instance)` will evaluate to false, then we'll try to do `instance.Skip(start)` and `Skip` extension will throw for the null `instance`.

File: src/Uno.UI.Runtime.Skia.Wpf/Extensions/UI/Xaml/Controls/TextBoxViewExtension.cs [hunk 7]
```diff
@@ -160,7 +164,19 @@ public void SetIsPassword(bool isPassword)
...
+		public void Select(int start, int length)
+		{
+			_currentInputWidget?.Select(start, length);
+
+#if __SKIA__
+			if (_currentInputWidget == null)
+			{
+				this.StartEntry();
+
+				_currentInputWidget?.Select(start, length);
```
Comment (Youssef1313) [inline src/Uno.UI.Runtime.Skia.Wpf/Extensions/UI/Xaml/Controls/TextBoxViewExtension.cs:176]: `StartEntry()` call should guarantee a non-null `_currentInputWidget` I think.

```suggestion
				_currentInputWidget!.Select(start, length);
```

File: src/Uno.UI.Runtime.Skia.Wpf/Extensions/UI/Xaml/Controls/TextBoxViewExtension.cs [hunk 8]
```diff
@@ -160,7 +164,19 @@ public void SetIsPassword(bool isPassword)
 			// No support for now.
 		}
 
-		public void Select(int start, int length) => _currentInputWidget?.Select(start, length);
+		public void Select(int start, int length)
+		{
+			_currentInputWidget?.Select(start, length);
+
+#if __SKIA__
```
Comment (Youssef1313) [inline src/Uno.UI.Runtime.Skia.Wpf/Extensions/UI/Xaml/Controls/TextBoxViewExtension.cs:171]: When can this be `false`?

File: src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.cs [hunk 9]
```diff
@@ -264,8 +264,24 @@ private void RaiseTextChanged()
 					_isTextChangedPending = false;
 				}
 			}
-
+			
 			_textBoxView?.SetTextNative(Text);
+		
+		}
+
+		private void ResetCarret()
```
Comment (Youssef1313) [inline src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.cs:272]: Minor typo :)

```suggestion
		private void ResetCaret()
```

File: src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.cs [hunk 10]
```diff
@@ -264,8 +264,24 @@ private void RaiseTextChanged()
...
+			this.SelectionStart = 0;
+			this.SelectionLength = 0;
+
+			if (_textBoxView != null)
+			{
+				_textBoxView.SelectionStart = 0;
+				_textBoxView.SelectionEnd = 0;
+			
+			}
+#endif
```
Comment (Youssef1313) [inline src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.cs:284]: I think this method is better in `TextBox.wasm.cs`

File: src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.wasm.cs [hunk 11]
```diff
@@ -44,6 +44,18 @@ private void UpdateTextBoxView()
 			}
 		}
 
+		partial void ValidateTextRange()
+		{
+			if (Text is { Length: > 0 } text || Text.TryGetStringByRange(SelectionStart, SelectionLength, out _))
```
Comment (Youssef1313) [inline src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.wasm.cs:49]: This is the only call of `TryGetStringByRange` and it doesn't use the `out` parameter. I think the extension does more than what it needs to do. The logic could be much simpler and can even be inlined here?

Comment (aesteves900) [reply src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.wasm.cs:49]: It was Carl's suggestion https://github.com/unoplatform/uno/pull/8265#discussion_r822025400

Comment (Youssef1313) [reply src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.wasm.cs:49]: Looking through PR comments, it seems my suggestion is exactly what @carldebilly is expecting in this comment: https://github.com/unoplatform/uno/pull/8265#discussion_r822016197

The comment you referenced (while is about this block of code) is unrelated to the concern of the unnecessary work.

File: src/Uno.UI.Runtime.Skia.Wpf/Extensions/UI/Xaml/Controls/TextBoxViewExtension.cs [hunk 12]
```diff
@@ -38,7 +38,11 @@ public void StartEntry()
 			_contentElement = textBox.ContentElement;
 
 			EnsureWidgetForAcceptsReturn();
-			textInputLayer.Children.Add(_currentInputWidget!);
+
+			if (textInputLayer.Children.Count == 0)
+			{
+				textInputLayer.Children.Add(_currentInputWidget!);
+			}
```
Comment (Youssef1313) [inline src/Uno.UI.Runtime.Skia.Wpf/Extensions/UI/Xaml/Controls/TextBoxViewExtension.cs:45]: I don't understand this change. How is it related to resetting the caret? What behavioral change does it introduce, if any?

File: src/Uno.UI.Runtime.Skia.Wpf/Extensions/UI/Xaml/Controls/TextBoxViewExtension.cs [hunk 13]
```diff
@@ -160,7 +164,19 @@ public void SetIsPassword(bool isPassword)
...
+
+#if __SKIA__
+			if (_currentInputWidget == null)
+			{
+				this.StartEntry();
+
+				_currentInputWidget!.Select(start, length);
+			}
+#endif
+		}
```
Comment (Youssef1313) [inline src/Uno.UI.Runtime.Skia.Wpf/Extensions/UI/Xaml/Controls/TextBoxViewExtension.cs:179]: I think this would result in the same thing, but is more simple.

```suggestion
		public void Select(int start, int length)
		{
			if (_currentInputWidget == null)
			{
				this.StartEntry();
			}

			_currentInputWidget!.Select(start, length);
		}
```

File: src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.Android.cs [hunk 14]
```diff
@@ -169,6 +169,15 @@ private void OnImeOptionsChanged(ImeAction imeAction)
...
 						_textBoxView.RequestFocus();
+
+						var selectionStart = this.SelectionStart;
+
+						if (selectionStart == 0)
+						{
+							int cursorPosition = selectionStart + _textBoxView?.Text?.Length ?? 0;
+
+							this.Select(cursorPosition, 0);
+						}
```
Comment (Youssef1313) [inline src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.Android.cs:180]: This is very confusing. Why are we altering the selection here? I think Android didn't have a bug in the first place.

I'm not sure, but it seems you're changing Android to match Wasm behavior instead of the opposite? The platform that had a bug, IIRC, was Wasm only.

File: src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_TextBox.cs [hunk 15]
```diff
@@ -156,12 +152,8 @@ public async Task When_Calling_Select_With_Out_Of_Range_Length()
 
 			WindowHelper.WindowContent = textBox;
 			await WindowHelper.WaitForLoaded(textBox);
-
-#if __WASM__ // Wasm is behaving differently than UWP and other platforms. https://github.com/unoplatform/uno/issues/7016
-			Assert.AreEqual(10, textBox.SelectionStart);
-#else
-			Assert.AreEqual(0, textBox.SelectionStart);
-#endif
+			textBox.Focus(FocusState.Programmatic);
```
Comment (Youssef1313) [inline src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_TextBox.cs:155]: Focus shouldn't be needed for the test to pass, I think. But it would be good to verify on UWP.

File: src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_TextBox.cs [hunk 16]
```diff
@@ -156,12 +152,8 @@ public async Task When_Calling_Select_With_Out_Of_Range_Length()
...
+			Assert.AreEqual(textBox.Text.Length, textBox.SelectionStart);
```
Comment (Youssef1313) [inline src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_TextBox.cs:156]: This behavior is resetting the caret to the end (selection start is now the text length). It should be 0 instead.

File: src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.Android.cs (hunk 14 above)
Comment (aesteves900) [reply src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.Android.cs:180]: The test fail because the Android does not mimic the UWP behavior, then I fixed that.

File: src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_TextBox.cs (hunk 16 above)
Comment (aesteves900) [reply src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_TextBox.cs:156]: In UWP when you focus the caret goes to the end, I equalized the Platforms to UWP behavior.

File: src/Uno.UI.Runtime.Skia.Wpf/Extensions/UI/Xaml/Controls/TextBoxViewExtension.cs (hunk 13 above)
Comment (aesteves900) [reply src/Uno.UI.Runtime.Skia.Wpf/Extensions/UI/Xaml/Controls/TextBoxViewExtension.cs:179]: The code block separated by #if __SKIA__ causes additional behavior that just is necessary on SKIA platform to mimic UWP behavior.

File: src/Uno.Foundation/Uno.Core.Extensions/Uno.Core.Extensions/StringExtensions.cs [hunk 17]
```diff
@@ -34,6 +34,20 @@ internal static partial class StringExtensions
 		private static readonly Lazy<Regex> _newLineRegex = new Lazy<Regex>(() => new Regex(@"^", RegexOptions.Multiline));
 #endif
 
+		public static bool TryGetStringByRange(this string instance, int start, int len, out string result)
```
[the remark this answers is not in the document]
Comment (aesteves900) [reply src/Uno.Foundation/Uno.Core.Extensions/Uno.Core.Extensions/StringExtensions.cs:37]: The method was removed.

File: src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.wasm.cs [hunk 18]
```diff
@@ -44,6 +44,18 @@ private void UpdateTextBoxView()
 			}
 		}
 
+		partial void ValidateTextRange()
+		{
```
Comment (carldebilly) [inline src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.wasm.cs:48]: It's confusing to have a method called `Validate...()` raising events or having side effects like reseting the caret...

Comment (aesteves900) [reply src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.wasm.cs:48]: Thank you for your comment, what do you suggest to clarify? When is not a valid range,  SelectionChanged fired to do the normal flow and not create side-effects.

File: src/Uno.UI.Runtime.Skia.Wpf/Extensions/UI/Xaml/Controls/TextBoxViewExtension.cs (hunk 8 above)
Comment (jeromelaban) [reply src/Uno.UI.Runtime.Skia.Wpf/Extensions/UI/Xaml/Controls/TextBoxViewExtension.cs:171]: When can this be false? I don't think we're reusing this file anywhere but here. You'll probably be able to remote the #if.

**Review by jeromelaban — APPROVED** [review]

**Review by carldebilly — APPROVED** [review]

File: src/Uno.UI.Runtime.Skia.Wpf/Extensions/UI/Xaml/Controls/TextBoxViewExtension.cs [hunk 19]
```diff
@@ -160,7 +164,17 @@ public void SetIsPassword(bool isPassword)
...
+		public void Select(int start, int length)
+		{
+			_currentInputWidget?.Select(start, length);
+
+			if (_currentInputWidget == null)
+			{
+				this.StartEntry();
+
+				_currentInputWidget!.Select(start, length);
+			}
```
Comment (Youssef1313) [inline src/Uno.UI.Runtime.Skia.Wpf/Extensions/UI/Xaml/Controls/TextBoxViewExtension.cs:176]:
```suggestion
			if (_currentInputWidget == null)
			{
				this.StartEntry();
			}

			_currentInputWidget!.Select(start, length);
```

File: src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.wasm.cs [hunk 20]
```diff
@@ -44,6 +44,18 @@ private void UpdateTextBoxView()
 			}
 		}
 
+		partial void OnSelectionChangedPartial()
+		{
+			if (Text is { Length: > 0 } text || string.IsNullOrEmpty(Text) || SelectionStart + SelectionLength > Text.Length)
```
Comment (Youssef1313) [inline src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.wasm.cs:49]: > Text is { Length: > 0 } text || string.IsNullOrEmpty(Text)

I'm confused. Isn't this always true?

File: src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.cs [hunk 21]
```diff
@@ -1000,10 +1011,8 @@ private protected override void OnIsEnabledChanged(IsEnabledChangedEventArgs e)
 
 		partial void OnIsEnabledChangedPartial(IsEnabledChangedEventArgs e);
 
-		private bool ShouldFocusOnPointerPressed(PointerRoutedEventArgs args)
-			// For mouse and pen, the TextBox should focus on pointer press
-			// (and then capture pointer to make sure to handle the whol down->move->up sequence).
-			// For touch we wait for the release to focus (avoid flickering in case of cancel due to scroll for instance).
```
Comment (Youssef1313) [inline src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.cs:1006]: Part of the comment was deleted. Was this intended?

File: src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.wasm.cs [hunk 22]
```diff
@@ -44,6 +44,18 @@ private void UpdateTextBoxView()
 			}
 		}
 
+		partial void OnSelectionChangedPartial()
```
Comment (Youssef1313) [inline src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.wasm.cs:47]: Judging from the tests (and from your comments). Wasm behavior was already correct, and tests don't capture changes in it. Why do we need any changes here at all?

File: src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.wasm.cs [hunk 23]
```diff
@@ -44,6 +44,18 @@ private void UpdateTextBoxView()
...
+			if (Text is { Length: > 0 } text || string.IsNullOrEmpty(Text) || SelectionStart + SelectionLength > Text.Length)
+			{
+				SelectionChanged?.Invoke(this, new RoutedEventArgs(this));
+			}
+			else
+			{
+				this.ResetCaret();
+			}
+		}		
+
```
Comment (Youssef1313) [inline src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.wasm.cs:58]:
```suggestion
```

File: src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.wasm.cs [hunk 24]
```diff
@@ -59,6 +71,21 @@ private void UpdateTextBoxView()
...
+			this.SelectionLength = 0;
+
+			if (_textBoxView != null)
+			{
+				_textBoxView.SelectionStart = 0;
+				_textBoxView.SelectionEnd = 0;
+
+			}
+		}
+
```
Comment (Youssef1313) [inline src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.wasm.cs:88]:
```suggestion
```

File: src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.cs [hunk 25]
```diff
@@ -908,7 +910,16 @@ private void DeleteButtonClick()
...
+		internal void OnSelectionChanged()
+		{
+#if __WASM__
+			this.OnSelectionChangedPartial();
+#else
+			SelectionChanged?.Invoke(this, new RoutedEventArgs(this));
+#endif
+		}
+
+		partial void OnSelectionChangedPartial();
```
Comment (Youssef1313) [inline src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.cs:922]:
```suggestion
		internal void OnSelectionChanged() => SelectionChanged?.Invoke(this, new RoutedEventArgs(this));
```

File: src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.wasm.cs (hunk 20 above)
Comment (Youssef1313) [reply src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.wasm.cs:49]: Because this is always true, I'm suggesting in the below comments to revert those changes and keep the always-true branch in the language-agnostic TextBox.cs
